store/src/main/java/org/apache/rocketmq/store/timer/TimerWheel.java

```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    public void flush() {
+        ByteBuffer bf = localBuffer.get();
+        bf.position(0);
+        bf.limit(wheelLength);
+        mappedByteBuffer.position(0);
+        mappedByteBuffer.limit(wheelLength);
+        for (int i = 0; i < wheelLength; i++) {
+            if (bf.get(i) != mappedByteBuffer.get(i)) {
+                mappedByteBuffer.put(i, bf.get(i));
+            }
+        }
+        this.mappedByteBuffer.force();
+    }
+
+    public Slot getSlot(long timeMs) {
+        Slot slot = getRawSlot(timeMs);
+        if (slot.timeMs != timeMs / precisionMs * precisionMs) {
+            return new Slot(-1, -1, -1);
+        }
+        return slot;
+    }
+
+    //testable
+    public Slot getRawSlot(long timeMs) {
+        localBuffer.get().position(getSlotIndex(timeMs) * Slot.SIZE);
+        return new Slot(localBuffer.get().getLong() * precisionMs,
+            localBuffer.get().getLong(), localBuffer.get().getLong(), localBuffer.get().getInt(), localBuffer.get().getInt());
+    }
+
+    public int getSlotIndex(long timeMs) {
+        return (int) (timeMs / precisionMs % (slotsTotal * 2));
+    }
+
+    public void putSlot(long timeMs, long firstPos, long lastPos) {
+        localBuffer.get().position(getSlotIndex(timeMs) * Slot.SIZE);
+        // To be compatible with previous version.
+        // The previous version's precision is fixed at 1000ms and it store timeMs / 1000 in slot.
+        localBuffer.get().putLong(timeMs / precisionMs);
+        localBuffer.get().putLong(firstPos);
+        localBuffer.get().putLong(lastPos);
+    }
+    public void putSlot(long timeMs, long firstPos, long lastPos, int num, int magic) {
+        localBuffer.get().position(getSlotIndex(timeMs) * Slot.SIZE);
+        localBuffer.get().putLong(timeMs / precisionMs);
+        localBuffer.get().putLong(firstPos);
+        localBuffer.get().putLong(lastPos);
+        localBuffer.get().putInt(num);
+        localBuffer.get().putInt(magic);
+    }
+
+    public void reviseSlot(long timeMs, long firstPos, long lastPos, boolean force) {
+        localBuffer.get().position(getSlotIndex(timeMs) * Slot.SIZE);
+
+        if (timeMs / precisionMs != localBuffer.get().getLong()) {
+            if (force) {
+                putSlot(timeMs, firstPos != IGNORE ? firstPos : lastPos, lastPos);
+            }
+        } else {
+            if (IGNORE != firstPos) {
+                localBuffer.get().putLong(firstPos);
+            } else {
+                localBuffer.get().getLong();
+            }
+            if (IGNORE != lastPos) {
+                localBuffer.get().putLong(lastPos);
+            }
+        }
+    }
+
+    //check the timerwheel to see if its stored offset > maxOffset in timerlog
+    public long checkPhyPos(long timeStartMs, long maxOffset) {
+        long minFirst = Long.MAX_VALUE;
+        int firstSlotIndex = getSlotIndex(timeStartMs);
+        for (int i = 0; i < slotsTotal * 2; i++) {
```

Comment:
I noticed that `slotsTotal * 2` appears multiple times. Why do we need to multiply it by 2? I don't understand.